src/collapse/dtrace.rs

```diff
@@ -0,0 +1,163 @@
+///
+/// To construct one, either use `dtrace::Folder::default()` or create an [`Options`] and use
+/// `dtrace::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// All lines until the next empty line are stack lines.
+    in_event: bool,
+
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+    count: usize,
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    _opt: Options, // to keep the interface consistant
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        for _ in 0..HEADER_LINES {
+            reader.read_line(&mut line)?;
+        }
+
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else {
+                self.in_event = true;
+                self.on_stack_line(line);
+            }
+        }
+        self.after_event();
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Self {
+            in_event: false,
+            stack: VecDeque::default(),
+            count: 0,
+            occurrences: HashMap::default(),
+            _opt: opt,
+        }
+    }
+}
+
+impl Folder {
+    fn declutter(frame: &str) -> Vec<String> {
+        let mut frame = CPP_REGEX.replace_all(frame, "$n").to_string();
+        if frame.is_empty() {
+            frame = "-".to_string();
+        };
+        let mut res = Vec::new();
+        for func in frame.split("->") {
+            let mut func = func.replace(';', ":").trim_start_matches('L').to_string();
+            if !res.is_empty() {
+                func.push_str("_[i]")
+            };
+            res.push(func);
+        }
+        res
+    }
+    fn stack_line_parts(line: &str) -> Option<(Vec<String>, &str)> {
+        let mut line = line.rsplitn(2, '+');
+        let second = line.next()?;
+        if let Some(path) = line.next() {
+            Some((Self::declutter(path), second))
+        } else {
+            Some((Self::declutter(second), ""))
+        }
+    }
+
+    // we have a stack line that shows one stack entry from the preceeding event, like:
+    //
+    //     unix`tsc_gethrtimeunscaled+0x21
+    //     genunix`gethrtime_unscaled+0xa
+    //     genunix`syscall_mstate+0x5d
+    //     unix`sys_syscall+0x10e
+    //       1
+    fn on_stack_line(&mut self, line: &str) {
+        let line = line.trim_start();
+        if let Ok(count) = line.parse::<usize>() {
+            self.count = count;
+        } else {
+            if let Some((mut funcs, _addr)) = Folder::stack_line_parts(line) {
```

Comment:
This still _always_ ignores the `addr`, which I don't think is what [upstream does](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L85).

Reply:
It implicitly does [upstream](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L54)  just with a constant to do so.

Reply:
added the logic for offset inclusion from upstream